File: examples/jev_triage_agent/agent.mjs
/** Scripted support-triage command agent using the official TypeSafe JavaScript SDK. */

import { createHash } from "node:crypto";
import { once } from "node:events";
import { readFile } from "node:fs/promises";
import { spawn } from "node:child_process";
import { createInterface } from "node:readline";
import { pathToFileURL } from "node:url";

import { TypeSafeClient, TypeSafeError, choice, noul, score } from "@typesafe-ai/sdk";

const QUESTIONS = {
  department: choice("Which team should handle this ticket?", {
    billing: "Payments, invoicing, and refunds",
    technical: "Bugs, outages, and integrations",
    sales: "Pricing, plans, and account questions",
  }),
  refund_requested: noul("The customer explicitly asks for a refund"),
  policy_supports: noul("The refund policy covers this situation", {
    true: "Duplicate charges are refundable",
    false: "Not covered",
  }),
  frustration: score("How frustrated is the customer?", [
    "Calm",
    "Frustrated but civil",
    "Very angry",
  ]),
};

class McpRpcError extends Error {}

function object(value, label) {
  if (value === null || Array.isArray(value) || typeof value !== "object") {
    throw new TypeError(`${label} must be a JSON object`);
  }
  return value;
}

function number(value, label) {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new TypeError(`${label} must be a number`);
  }
  return value;
}

/** Minimal MT19937 implementation matching `random.Random(str).randrange()` in agent.py. */
class PythonRandom {
  constructor(seed) {
    const bytes = Buffer.from(seed, "utf8");
    const digest = createHash("sha512").update(bytes).digest();
    let value = BigInt(`0x${Buffer.concat([bytes, digest]).toString("hex")}`);
    const key = [];
    while (value > 0n) {
      key.push(Number(value & 0xffffffffn));
      value >>= 32n;
    }
    this.mt = new Uint32Array(624);
    this.index = 624;
    this.#initByArray(key);
  }

  #init(seed) {
    this.mt[0] = seed >>> 0;
    for (let index = 1; index < 624; index += 1) {
      const previous = this.mt[index - 1];
      this.mt[index] = (Math.imul(1812433253, previous ^ (previous >>> 30)) + index) >>> 0;
    }
  }

  #initByArray(key) {
    this.#init(19650218);
    let i = 1;
    let j = 0;
    for (let count = Math.max(624, key.length); count > 0; count -= 1) {
      const previous = this.mt[i - 1];
      this.mt[i] =
        ((this.mt[i] ^ Math.imul(previous ^ (previous >>> 30), 1664525)) + key[j] + j) >>> 0;
      i += 1;
      j += 1;
      if (i >= 624) {
        this.mt[0] = this.mt[623];
        i = 1;
      }
      if (j >= key.length) j = 0;
    }
    for (let count = 623; count > 0; count -= 1) {
      const previous = this.mt[i - 1];
      this.mt[i] =
        ((this.mt[i] ^ Math.imul(previous ^ (previous >>> 30), 1566083941)) - i) >>> 0;
      i += 1;
      if (i >= 624) {
        this.mt[0] = this.mt[623];
        i = 1;
      }
    }
    this.mt[0] = 0x80000000;
  }

  #uint32() {
    if (this.index >= 624) {
      for (let index = 0; index < 624; index += 1) {
        const value = (this.mt[index] & 0x80000000) | (this.mt[(index + 1) % 624] & 0x7fffffff);
        this.mt[index] =
          this.mt[(index + 397) % 624] ^
          (value >>> 1) ^
          (value & 1 ? 0x9908b0df : 0);
      }
      this.index = 0;
    }
    let value = this.mt[this.index];
    this.index += 1;
    value ^= value >>> 11;
    value ^= (value << 7) & 0x9d2c5680;
    value ^= (value << 15) & 0xefc60000;
    value ^= value >>> 18;
    return value >>> 0;
  }

  randrange(stop) {
    const bits = Math.floor(Math.log2(stop)) + 1;
    let value;
    do {
      value = this.#uint32() >>> (32 - bits);
    } while (value >= stop);
    return value;
  }
}

export function ticketIndexForSeed(seed, count) {
  return new PythonRandom(`${seed}:triage-ticket`).randrange(count);
}

function selectedTicket(task, seed) {
  const tickets = task.tickets;
  if (!Array.isArray(tickets) || tickets.length === 0) {
    throw new TypeError("task.tickets must be a non-empty list");
  }
  return object(tickets[ticketIndexForSeed(seed, tickets.length)], "selected ticket");
}

class McpSession {
  static async open(configPath) {
    const config = object(JSON.parse(await readFile(configPath, "utf8")), "MCP config");
    const servers = object(config.mcpServers, "mcpServers");
    const entries = Object.values(servers);
    if (entries.length !== 1) throw new Error("MCP config must contain exactly one server");
    const entry = object(entries[0], "MCP server entry");
    if (typeof entry.command !== "string") {
      throw new TypeError("MCP server command must be a string");
    }
    const args = entry.args ?? [];
    const extraEnv = object(entry.env ?? {}, "MCP server env");
    if (!Array.isArray(args) || args.some((item) => typeof item !== "string")) {
      throw new TypeError("MCP server args must be strings");
    }
    if (Object.values(extraEnv).some((value) => typeof value !== "string")) {
      throw new TypeError("MCP server env values must be strings");
    }
    const child = spawn(entry.command, args, {
      env: { ...process.env, ...extraEnv },
      stdio: ["pipe", "pipe", "ignore"],
    });
    const session = new McpSession(child);
    try {
      await session.rpc("initialize", {
        protocolVersion: "2026-07-28",
        capabilities: {},
        clientInfo: { name: "arci-jev-triage-node", version: "0.6" },
      });
      await session.send({ jsonrpc: "2.0", method: "notifications/initialized" });
      await session.rpc("tools/list", {});
      return session;
    } catch (error) {
      await session.close();
      throw error;
    }
  }

  constructor(child) {
    this.child = child;
    this.exited = once(child, "exit");
    this.lines = createInterface({ input: child.stdout, crlfDelay: Infinity })[Symbol.asyncIterator]();
    this.nextId = 0;
  }

  async send(message) {
    const line = `${JSON.stringify(message)}\n`;
    if (!this.child.stdin.write(line, "utf8")) await once(this.child.stdin, "drain");
  }

  async rpc(method, params) {
    this.nextId += 1;
    const id = `req-${this.nextId}`;
    await this.send({ jsonrpc: "2.0", id, method, params });
    while (true) {
      const line = await this.lines.next();
      if (line.done) throw new Error("MCP server exited before replying");
      const reply = object(JSON.parse(line.value), "MCP reply");
      if (reply.id !== id) continue;
      if ("error" in reply) throw new McpRpcError("MCP request failed");
      return object(reply.result, "MCP result");
    }
  }

  async call(name, args) {
    const result = await this.rpc("tools/call", { name, arguments: args });
    if (result.isError === true) throw new McpRpcError(`tool ${name} failed`);
  }

  async close() {
    if (!this.child.stdin.destroyed) this.child.stdin.end();
    const timer = setTimeout(() => this.child.kill("SIGKILL"), 5000);
    try {
      await this.exited;
    } finally {
      clearTimeout(timer);
    }
  }
}

function decideAction(answers) {
  const department = object(answers.department, "department answer");
  if (department.choice === "billing") {
    const refund = object(answers.refund_requested, "refund_requested answer");
    const policy = object(answers.policy_supports, "policy_supports answer");
    if (
      number(refund.noul, "refund_requested.noul") > 0.5 &&
      number(policy.noul, "policy_supports.noul") > 0.5
    ) {
      return "issue_refund";
    }
  }
  return "reply";
}

export function gatedAction(answers) {
  const department = object(answers.department, "department answer");
  const confidence = number(department.confidence, "department.confidence");
  if (confidence < 0.6) return null;
  const action = decideAction(answers);
  if (action === "issue_refund") {
    const refund = object(answers.refund_requested, "refund_requested answer");
    const policy = object(answers.policy_supports, "policy_supports answer");
    if (
      confidence < 0.85 ||
      Math.min(
        number(refund.noul, "refund_requested.noul"),
        number(policy.noul, "policy_supports.noul"),
      ) < 0.85
    ) {
      return null;
    }
  }
  return action;
}

async function act(configPath, ticket, action) {
  if (typeof ticket.ticket_id !== "string") throw new TypeError("ticket_id must be a string");
  const session = await McpSession.open(configPath);
  try {
    if (action === "issue_refund") {
      await session.call("issue_refund", { ticket_id: ticket.ticket_id });
    } else if (action === "escalate") {
      await session.call("escalate", {
        ticket_id: ticket.ticket_id,
        reason: "decision confidence or provider failure",
      });
    } else {
      await session.call("reply", {
        ticket_id: ticket.ticket_id,
        message: "Your ticket was routed to the right team.",
      });
    }
  } finally {
    await session.close();
  }
}

function parseArgs(argv) {
  const values = {};
  for (let index = 0; index < argv.length; index += 2) {
    const flag = argv[index];
    const value = argv[index + 1];
    if (!["--mcp-config", "--task-file", "--variant"].includes(flag) || value === undefined) {
      throw new Error("usage: agent.mjs --mcp-config PATH --task-file PATH --variant a|b|c");
    }
    values[flag] = value;
  }
  if (!values["--mcp-config"] || !values["--task-file"] || !["a", "b", "c"].includes(values["--variant"])) {
    throw new Error("usage: agent.mjs --mcp-config PATH --task-file PATH --variant a|b|c");
  }
  return values;
}

export async function main() {
  const args = parseArgs(process.argv.slice(2));
  const task = object(JSON.parse(await readFile(args["--task-file"], "utf8")), "task");
  const seed = Number.parseInt(process.env.ARCI_SEED ?? "", 10);
  if (!Number.isSafeInteger(seed)) throw new TypeError("ARCI_SEED must be an integer");
  const ticket = selectedTicket(task, seed);
  let action;
  try {
    const client = new TypeSafeClient();
    const response = await client.systemOne({
      state: { ticket },
      questions: QUESTIONS,
      model: "jev-latest",
    });
    action = gatedAction(response.answers);
  } catch (error) {
    if (!(error instanceof TypeSafeError)) throw error;
    action = null;
  }
  if (action === null) {
    if (args["--variant"] === "b") return;
    action = "escalate";
  }
  await act(args["--mcp-config"], ticket, action);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((error) => {
    console.error(error instanceof Error ? `${error.name}: ${error.message}` : "agent failed");
    process.exitCode = 1;
  });
}

File: examples/jev_triage_agent/agent.py
"""Scripted support-triage command agent using HTTP decisions and MCP actions."""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import urllib.error
import urllib.request
from collections.abc import Sequence
from contextlib import suppress
from pathlib import Path
from typing import Any, Literal, TypeAlias, cast

from .world import JsonValue, ticket_index_for_seed

JsonObject: TypeAlias = dict[str, JsonValue]
Action: TypeAlias = Literal["issue_refund", "reply"]

QUESTIONS: JsonObject = {
    "department": {
        "type": "choice",
        "instructions": "Which team should handle this ticket?",
        "criteria": {
            "billing": "Payments, invoicing, and refunds",
            "technical": "Bugs, outages, and integrations",
            "sales": "Pricing, plans, and account questions",
        },
    },
    "refund_requested": {
        "type": "noul",
        "instructions": "The customer explicitly asks for a refund",
    },
    "policy_supports": {
        "type": "noul",
        "instructions": "The refund policy covers this situation",
        "criteria": {"true": "Duplicate charges are refundable", "false": "Not covered"},
    },
    "frustration": {
        "type": "score",
        "instructions": "How frustrated is the customer?",
        "criteria": ["Calm", "Frustrated but civil", "Very angry"],
    },
}


class DecisionProviderError(RuntimeError):
    """The decision endpoint could not produce a usable response."""


class McpRpcError(RuntimeError):
    """The MCP server rejected a request."""


class _RejectRedirects(urllib.request.HTTPRedirectHandler):
    def redirect_request(
        self,
        req: urllib.request.Request,
        fp: Any,
        code: int,
        msg: str,
        headers: Any,
        newurl: str,
    ) -> None:
        del fp, newurl
        raise urllib.error.HTTPError(req.full_url, code, msg, headers, None)


def _object(value: object, label: str) -> JsonObject:
    if not isinstance(value, dict) or any(not isinstance(key, str) for key in value):
        raise TypeError(f"{label} must be a JSON object")
    return cast(JsonObject, value)


def _number(value: JsonValue | None, label: str) -> float:
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise TypeError(f"{label} must be a number")
    return float(value)


def _ticket(task: JsonObject, seed: int) -> JsonObject:
    tickets = task.get("tickets")
    if not isinstance(tickets, list) or not tickets:
        raise TypeError("task.tickets must be a non-empty list")
    return _object(tickets[ticket_index_for_seed(seed)], "selected ticket")


class Endpoint:
    def __init__(self) -> None:
        self._base = os.environ["TYPESAFE_BASE_URL"].rstrip("/")
        self._key = os.environ["TYPESAFE_API_KEY"]
        self._opener = urllib.request.build_opener(
            urllib.request.ProxyHandler({}), _RejectRedirects()
        )

    def _post(self, body: JsonObject) -> tuple[int, JsonValue]:
        data = json.dumps(body, ensure_ascii=False, allow_nan=False).encode("utf-8")
        request = urllib.request.Request(
            self._base + "/v1/systemone",
            data=data,
            method="POST",
            headers={
                "Authorization": f"Bearer {self._key}",
                "Content-Type": "application/json",
            },
        )
        try:
            with self._opener.open(request, timeout=15.0) as response:
                raw = response.read().decode("utf-8")
                return response.status, cast(JsonValue, json.loads(raw))
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            try:
                return exc.code, cast(JsonValue, json.loads(raw))
            except json.JSONDecodeError:
                return exc.code, raw
        except (OSError, TimeoutError, UnicodeError, json.JSONDecodeError) as exc:
            raise DecisionProviderError(type(exc).__name__) from exc

    def ask(self, ticket: JsonObject) -> JsonObject:
        request: JsonObject = {
            "state": {"ticket": ticket},
            "model": "jev-latest",
            "questions": QUESTIONS,
        }
        status = 0
        payload: JsonValue = None
        for _ in range(3):
            status, payload = self._post(request)
            if status != 529:
                break
        if status != 200:
            raise DecisionProviderError(f"decision endpoint returned HTTP {status}")
        return _object(payload, "decision response")


class McpSession:
    def __init__(self, config_path: str | Path) -> None:
        config = _object(json.loads(Path(config_path).read_text(encoding="utf-8")), "MCP config")
        servers = _object(config.get("mcpServers"), "mcpServers")
        if len(servers) != 1:
            raise ValueError("MCP config must contain exactly one server")
        entry = _object(next(iter(servers.values())), "MCP server entry")
        command = entry.get("command")
        raw_args = entry.get("args", [])
        raw_env = entry.get("env", {})
        if not isinstance(command, str):
            raise TypeError("MCP server command must be a string")
        if not isinstance(raw_args, list) or any(not isinstance(item, str) for item in raw_args):
            raise TypeError("MCP server args must be strings")
        env = _object(raw_env, "MCP server env")
        if any(not isinstance(value, str) for value in env.values()):
            raise TypeError("MCP server env values must be strings")
        self._process = subprocess.Popen(
            [command, *cast(list[str], raw_args)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            encoding="utf-8",
            env={**os.environ, **cast(dict[str, str], env)},
        )
        self._next_id = 0
        try:
            self._rpc(
                "initialize",
                {
                    "protocolVersion": "2026-07-28",
                    "capabilities": {},
                    "clientInfo": {"name": "arci-jev-triage", "version": "0.5"},
                },
            )
            self._send({"jsonrpc": "2.0", "method": "notifications/initialized"})
        except BaseException:
            self.close()
            raise

    def _send(self, message: JsonObject) -> None:
        stream = self._process.stdin
        if stream is None:
            raise RuntimeError("MCP server stdin is unavailable")
        stream.write(json.dumps(message, ensure_ascii=False, allow_nan=False) + "\n")
        stream.flush()

    def _rpc(self, method: str, params: JsonObject) -> JsonObject:
        self._next_id += 1
        request_id = f"req-{self._next_id}"
        self._send({"jsonrpc": "2.0", "id": request_id, "method": method, "params": params})
        stream = self._process.stdout
        if stream is None:
            raise RuntimeError("MCP server stdout is unavailable")
        while True:
            line = stream.readline()
            if not line:
                raise RuntimeError("MCP server exited before replying")
            reply = _object(json.loads(line), "MCP reply")
            if reply.get("id") != request_id:
                continue
            if "error" in reply:
                raise McpRpcError("MCP request failed")
            return _object(reply.get("result"), "MCP result")

    def call(self, name: str, arguments: JsonObject) -> None:
        result = self._rpc("tools/call", {"name": name, "arguments": arguments})
        if result.get("isError") is True:
            raise McpRpcError(f"tool {name} failed")

    def close(self) -> None:
        if self._process.stdin is not None and not self._process.stdin.closed:
            with suppress(OSError):
                self._process.stdin.close()
        try:
            self._process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self._process.kill()
            self._process.wait(timeout=5)
        finally:
            if self._process.stdout is not None:
                self._process.stdout.close()


def decide_action(answers: JsonObject) -> Action:
    department = _object(answers.get("department"), "department answer")
    choice = department.get("choice")
    if choice == "billing":
        refund = _object(answers.get("refund_requested"), "refund_requested answer")
        policy = _object(answers.get("policy_supports"), "policy_supports answer")
        if (
            _number(refund.get("noul"), "refund_requested.noul") > 0.5
            and _number(policy.get("noul"), "policy_supports.noul") > 0.5
        ):
            return "issue_refund"
    return "reply"


def gated_action(answers: JsonObject) -> Action | None:
    """Use the documented 0.6 floor and 0.85 threshold for a refund."""
    department = _object(answers.get("department"), "department answer")
    confidence = _number(department.get("confidence"), "department.confidence")
    if confidence < 0.6:
        return None
    action = decide_action(answers)
    if action == "issue_refund":
        refund = _object(answers.get("refund_requested"), "refund_requested answer")
        policy = _object(answers.get("policy_supports"), "policy_supports answer")
        if (
            confidence < 0.85
            or min(
                _number(refund.get("noul"), "refund_requested.noul"),
                _number(policy.get("noul"), "policy_supports.noul"),
            )
            < 0.85
        ):
            return None
    return action


def _act(config_path: str, ticket: JsonObject, action: Action | Literal["escalate"]) -> None:
    ticket_id = ticket.get("ticket_id")
    if not isinstance(ticket_id, str):
        raise TypeError("ticket_id must be a string")
    session = McpSession(config_path)
    try:
        if action == "issue_refund":
            session.call("issue_refund", {"ticket_id": ticket_id})
        elif action == "escalate":
            session.call(
                "escalate",
                {"ticket_id": ticket_id, "reason": "decision confidence or provider failure"},
            )
        else:
            session.call(
                "reply",
                {"ticket_id": ticket_id, "message": "Your ticket was routed to the right team."},
            )
    finally:
        session.close()


def _parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument("--mcp-config", required=True)
    parser.add_argument("--task-file", required=True)
    parser.add_argument("--variant", choices=("a", "b", "c"), required=True)
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    args = _parser().parse_args(argv)
    task = _object(json.loads(Path(cast(str, args.task_file)).read_text(encoding="utf-8")), "task")
    seed = int(os.environ["ARCI_SEED"])
    ticket = _ticket(task, seed)
    variant = cast(str, args.variant)
    try:
        response = Endpoint().ask(ticket)
        answers = _object(response.get("answers"), "answers")
        action: Action | Literal["escalate"] | None = gated_action(answers)
    except DecisionProviderError:
        action = None
    if action is None:
        if variant == "b":
            return 0
        action = "escalate"
    _act(cast(str, args.mcp_config), ticket, action)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
